### ISSUE TITLE
Ancient Map Upgrades

### ISSUE THREAD

**Issue by esampson** [Mindwerks/worldengine#177]
I'm starting some experiments with upgrading the ancient map routines. Just as a proof of concept I modified the mountains.
![ancient_map_ancient_test-old](https://cloud.githubusercontent.com/assets/8743461/10902083/c2d611bc-81ad-11e5-99a9-6449091c1dc7.png)
What I'm doing is basically just a replacement of the draw_mountain routine. Instead of the existing routine I am calling a new routine called '_stamp' that is fed a bitmap. The current setup is extremely rudimentary. I have 4 different PNG files that are loaded and I randomly feed one of them to the stamp routine.

So we could probably replace all of the individual _draw_(something) routines with the _stamp routine when an object is being draw, which would probably be good. I've also got an idea for handling areas that are textures such as ice and steppes so we could replace those remaining routines with this second routine.

One of the questions I have is whether we want to have the final system load bitmap files or whether we want to embed the data in the program itself using techniques such as PIL's image.fromstring() function. Loading individual PNG bitmaps would certainly be easier, program-wise, but it would also probably result in a lot of files being added as well.

**Comment by ftomassetti** [comment]
Not bad, not bad at all...

My first comment is that the mountain chosen should not be completely random but instead depend on the elevation (higher elevation, higher image). Perhaps you could dynamically rescale the images also.

I would be perfectly fine to add a lot of files to the project but on this it is better to hear the opinion of @psi29a who did a lot of work to clean the repository after I made it grow to a few hundreds megabytes...

**Comment by esampson** [comment]
Actually if you look carefully you can see that it does rescale the images. There are some smaller mountains.

As for doing things like using certain mountains when the elevation is higher, that would be extremely easy. All we do is go into the routine that is placing all the mountains and rather than simply choosing one at random we first check elevation. If it is over a certain amount we randomly choose from a set of images for tall mountains, if it is between a certain range we choose from a set of regular mountains and if it is below that range we choose from a set of hills.

Since we don't have to write a new routine for each new type of image it is about 50 times easier to add new images then it was before. Additionally, if we continue to load PNG images for our icons that should put us in a very good position for letting other people 'reskin' the ancient map routine with their own sets of icons. At its simplest we just have an extra command line option for drawing ancient maps that specifies an alternate directory for the icons.

**Comment by ftomassetti** [comment]
"Actually if you look carefully you can see that it does rescale the images. There are some smaller mountains."

I did not know if they were just different images :)

I like the idea of having different skins....

**Comment by tcld** [comment]
I actually like the somewhat pictogram-like nature of the ancient map (and as far as I see that is the only map that is affected by this). Things are very clear to the eyes.

That doesn't mean that I don't like the idea, though. Just that the images shouldn't be too detailed or the map will be hard to parse. (But I don't really like the idea of adding a huge amount of images to the repository. There is probably a better way.)

On a side note, I also think that there is a lot of room for improvement in ```drawing_functions.py```; a lot of code seems to be repeated, many functions could be condensed into one (simple) function etc. So if you are working in there, you might want to think about some simplifications that can be made.

**Comment by ftomassetti** [comment]
Yeah, the code drawing ancient maps was written as a sort of hack but I have to say it is one of the features I like the most about WorldEngine :)

**Comment by esampson** [comment]
ftomasseti: Yeah, this was mostly just a 'proof of concept' so I could look for any potential problems and so I could make sure it could be done without a complete rewrite of the existing routines. All I did was replace the older mountain icon with these new bitmapped images.

tcld: Like I said above, this was mostly proof of concept at the moment. We probably couldn't even use the specific images I used as I sort of ripped them from Campaign Cartographer 3. I find the new mountains too 'mushy' as well as they sort of seem to be bleeding together, but we could even use images that are almost identical to the existing images, if we want.

There seems to be pluses and minuses to some sort of simple method of storing bitmaps which is why I can't really decide what method I like best (and why I'm asking everyone else what they think would work). What about basically just zipping the entire directory of icons into a single file and then reading from that. I believe we would need to add some routines to pull specific files from a zip but it would probably cut the size of the file(s) considerably while remaining fairly easy for other people to author new icon sets.

**Comment by tcld** [comment]
Another approach might be to have all the different mountains tiled in a single PNG.

**Comment by esampson** [comment]
Possibly. I think if we are going that route we might want to put all icons in a single PNG and then have a text file that tells the software how to 'cut up' the image. This way people making alternate icon sets wouldn't have to have 4 mountain icons, no more and no less.

**Comment by tcld** [comment]
So that would make two files, with the PNG being separable in whatever way the txt-provider wants it. Sounds very good (and extendable) to me, to be honest.

**Comment by esampson** [comment]
Yeah. Each skin would be two files; a PNG and a text. The PNG would contain all the image data while the text file would identify the different image components and contain a few extra bits of information such as scaling data and offsets.

The biggest concern I have is that it would require authors to specify the bounding box for every icon, which could be a bit tedious and might have to be redone if the image needs to be edited for some reason. A zip file containing individual files would be much easier for an author to create but we would have to add a library to handle the zip files.

**Comment by ftomassetti** [comment]
For inspiring us: http://www.fantasticmaps.com/2015/01/how-to-draw-shade-and-colour-a-mountain-range/

**Comment by rbb** [comment]
For  a zip vs a single file...

FYI, there are tools like [Tiled](http://sourceforge.net/projects/tiled/) for creating sprite sheets, and tools like [TMX loader](http://www.pygame.org/project-Tiled+TMX+Loader-2036-3633.html), in pygame that assist with loading the sprite sheets.

Maybe this info helps?

**Comment by ftomassetti** [comment]
it would very interesting to use Tiled to edit worlds generated by WorldEngine... perhaps we should discuss this possibility in the manual

**Comment by ftomassetti** [comment]
@rbb thanks for talking about Tiled, now I am working on the integration with it in #180

**Comment by rbb** [comment]
@ftomassetti: Glad it was helpful

However, I was thinking about using Tiled in a different way - for creating skins for the ancient_maps routines to use.

**Comment by esampson** [comment]
More experimentation:

![ancient_map_seed_2](https://cloud.githubusercontent.com/assets/8743461/10926403/f2c32baa-824b-11e5-8456-0f5ed2e58d27.png)

**Comment by esampson** [comment]
Combined most of the 'icon' passes into a single pass. Since it goes from top to bottom this eliminates certain bits of weirdness I was getting with tree trunks showing up at the edge of one type of forest when it met another type of forest. I think it also improves speed marginally.

I won't need to be making any use of Tiled in the software I'm working on. Cutting out the different sections based on coordinates from a text file will be pretty trivial. I've still got concerns over it adding an extra complication for authors, but that's a separate issue.

Next up on my agenda; creating textures from loaded bitmaps and then applying them based on masks. This will allow bitmaps to be used for things such as the ice fields, savanna, and steppes instead of the current pattern. 

After that I have some ideas to test concerning rivers. I think that if I make a pass over the existing masks and have the rivers cut into them in a manner similar to how a placed icon erases part of the mask so that the next icon isn't immediately on top of it that I will be able to make it look more like the river is running through a forest/desert/jungle rather than on top of it.

Once that's in place I'll look at making the code read from a text file to get all the values (they're hardwired at the moment while I'm working) and I think this will be good to go.

**Comment by ftomassetti** [comment]
It looks nice indeed. About rivers it could be interesting to consider what happens when you use a resize factor: right now if I use a factor 4, it means that every river is 4 pixel thick, which is not necessarily really nice.

Another thing about the ancient map is that it looks very crowded. It was not something caused by you of course but perhaps it could be something to consider among the improvements needed.

**Comment by esampson** [comment]
Actually, crowding definitely is something that I can control. I just have to increase the various radius. I'll run off a new map in a little bit with increased radius on the different icons so you can see.

That's the thing that will be nice, though, about doing this. People who want those maps where trees are tightly clumped together can do it while those people who want them spaced further apart will be able to do that as well.

**Comment by ftomassetti** [comment]
Yes but I was not thinking about spacing trees representing a forest but having more open spaces where no forest is represented. I think that currently most of the planet is covered by different kinds of forest biome. However when we think about the typical old map we expect to see a few forests, a few mountains, a couple of rivers and lots of empty space I guess. I have very confused ideas about how to reconciliate that expectation with a realistic world, just sharing my thoughts :)

**Comment by rbb** [comment]
Just a nitpick here...

I notice that in the snowy areas, where mountains intrude into these areas, there isn't any snow on the mountains. Seems like there should be. Of course, there should probably be some mountains without snow too.

![63436596-827b-11e5-8ae1-ada3d04fb91c](https://cloud.githubusercontent.com/assets/975429/10943614/51216262-82d2-11e5-92ca-75c50faf0e58.png)

**Comment by ftomassetti** [comment]
this is a good point...

**Comment by esampson** [comment]
The main reason for no snow on the mountains is because the original ancient map code doesn't support that. It shouldn't be too difficult to add it, however.

**Comment by rbb** [comment]
In that case, maybe snowy mountains is something for the next release, so we can get this one out the door?

**Comment by esampson** [comment]
>Yes but I was not thinking about spacing trees representing a forest but having more open spaces >where no forest is represented. I think that currently most of the planet is covered by different kinds of >forest biome. However when we think about the typical old map we expect to see a few forests, a few >mountains, a couple of rivers and lots of empty space I guess. I have very confused ideas about how >to reconciliate that expectation with a realistic world, just sharing my thoughts :)

Actually, I think this map may have more open areas than you are thinking, possibly indicating that I need better textures on my open lands (my goal for today). The large western landmass does have a very substantial jungle but I think it's just sort of a perfect storm being located right along the equator and getting a large clump of rain.

The good news is that you could regenerate the world with some shifts to humidity and perhaps the temperature/precipitation curve to break those up.

**Comment by esampson** [comment]
>In that case, maybe snowy mountains is something for the next release, so we can get this one out the door?

Part of the reason I'm doing this as an issue rather than a pull request. We've got so much stuff that's been done that I didn't want to toss one more thing in there before our next 'release' (unless the release gets pushed way back). Most likely this entire set of routines won't be made available until some time after our next release.

**Comment by ftomassetti** [comment]
I think I am falling in loves with the new mountains. The way you used partial overlaps really make a huge difference.

**Comment by psi29a** [comment]
I really like them too...

**Comment by tcld** [comment]
@esampson : Is the code so complicated? Once you have a decent setup, having a couple more eyes to look over the code now and then might be useful. Even if the PR isn't supposed to be merged immediately.
That also helps to prevent mass-suicide in case you are grabbed by aliens.

Your decision, I guess.

**Comment by psi29a** [comment]
Please send a PR, you're not overloading anyone. ;)

They are allowed to queue up.

**Comment by esampson** [comment]
It's not that the code is complicated or anything. It's just that it isn't ready yet. I'm doing some things sort of back to front so that I can figure out the format for the text files by seeing what I have to do to make the various components work. Now that I'm finished getting the basic components together I can look at defining a class for style elements to make everything easier to handle, then working backwards from that make the file loading command. 

Speaking of which, there's a comment at the start of draw.py that mentions only working on pixels and not whole images. While I've been cobbling stuff together I've been taking advantage of PIL routines directly with the intention of then moving those calls over to drawing_routines.py when I finish up. Of course if perspectives have shifted and we no longer care about keeping PIL calls out of drawing.py I won't have to worry about that, so I just wanted to double check and make sure that's not a legacy concept we are no longer worrying about.

As for protecting against mass-suicide or people who want to peer at the ugly code while I'm still working things out, you should be able to see it in my fork on github, I believe. I'm still doing commits and the like. I just am not issuing a pull request at the moment. :)

Added in a routine today that lets me take a bitmap and use it as a pattern to fill a region. It required a slight rework of masks so at present there's two forms of masks being created. My current plan is to replace the older form being used by icons with the new form being used by patterns. This is because any climate zone will be able to be filled with icons or pattern when I finish. It will simply be one of the options that the style author defines.

Tomorrow my plans are to replace all the old mask routines with the new ones, look at a way to have snow covered mountains, and possibly look at having the rivers carve through the icon filled terrains better. In the meantime, here's the latest output.

![ancient_map_seed_2](https://cloud.githubusercontent.com/assets/8743461/10961653/e9835aba-8345-11e5-81a2-aa82eb54d253.png)

**Comment by tcld** [comment]
When you mentioned PIL I realized that your work here and the stuff I did in #163 would very much be incompatible. :( I, too, created new routines to read/write images, and I removed Pillow for it.
My PR might not be merged, so it's probably me who would have to rework a lot of code. But since you want to take your time, I can't be certain.

I will take a look at your work later, and if you wanted to, you could try to maybe draw some inspiration from my PRs image_io.py.

**Comment by psi29a** [comment]
What you can do is make a PR with the title: DO NOT MERGE, WIP

That way you can take advantage of Travis-CI, Appeyor and of course people here who want to help! :D

You can of course turn on Travis-CI for your own fork too if you want.

**Comment by esampson** [comment]
![ancient_map_seed_2](https://cloud.githubusercontent.com/assets/8743461/11024569/0b8f73f8-8645-11e5-81b4-deb7b240f7b5.png)

**Comment by ftomassetti** [comment]
Very, very, very beatiful.

Two comments:
* the shadow is very beatiful but I would make it optional, as an alternative to the outer border
* perhaps rivers should be drawn before mountains, so that mountains partially cover them

**Comment by esampson** [comment]
Coastline coloring can definitely be made optional. I might also make the blurring factor controllable from the command line.

Placing the rivers before mountains might be a little problematic. In the case of the map above I think most of the rivers would disappear. Unfortunately cutting out the mountains around the rivers probably wouldn't work either. I'll have to continue experimenting until I can find a solution that I like.

Code is now loading only a single bitmap and cutting out all the separate images from it. Next up, reading from a text file rather than using hard coded values.

![ancient_map_seed_2](https://cloud.githubusercontent.com/assets/8743461/11039280/ba293b6c-86bb-11e5-84e7-cf7501ea4842.png)

![icons](https://cloud.githubusercontent.com/assets/8743461/11039287/c5a9d884-86bb-11e5-947d-1c63aa147595.PNG)

**Comment by psi29a** [comment]
Is there a way to make rivers less obtrusive, maybe with a transparency mask?

**Comment by esampson** [comment]
I could certainly do a transparency for the rivers but I'm not sure how much that would help. I've actually got some other things I want to try but they would probably make the rivers a little more obtrusive, although it should make them much better looking.

**Comment by rbb** [comment]
I know that it is a purely 2D drawing process, but since the icons look a little isometric, maybe drawing things in order from West to East, North to South (ie, like a TV scans), would help layer things in a way that would look good? I haven't looked at the code in a while, and it might be a pain to reorder the drawing. Just a thought.

**Comment by esampson** [comment]
That's actually something I've already implemented. Prior to working on this code different terrain types were rendered one at a time. Now all the textured areas are rendered first and then all the icon areas are rendered as a single pass. There was an interim stage where all the icons except for mountains were rendered in a single pass and then mountains were rendered separately. The current method hides a lot of the lower hills, but it produces a more aesthetically pleasing result.

**Comment by rbb** [comment]
Would there be a reasonable way to interleave the rendering of the textured and icon areas? Of course, its probably easier said than done?

**Comment by rbb** [comment]
Hmm. Maybe the rendering order should be dictated by the elevation map?

**Comment by esampson** [comment]
Not really because textured areas are done with a simply paste command using a mask. It's ok, however, because you really don't want to interleave them. Textured areas are for lands that don't have tall features, so things like steppes, savannas, and scrublands. 

Icons come in with their insert point on the center of their bottom row (I may actually change that in the future so that the base point it manually set). This means that they will overlap textures 'above' them but they won't overlap textures 'below'.

**Comment by esampson** [comment]
Hope that makes sense.

**Comment by rbb** [comment]
Yea, makes sense.

**Comment by tcld** [comment]
I have been browsing your *NewMap*-branch a bit. Didn't find a proper way to add comments to your code, so I'll just post some things here.

[draw.py - line 128](https://github.com/Mindwerks/worldengine/compare/master...esampson:NewMap#diff-993da1fbf1c1e611672d9ced4ed7b4a7R128)
I noticed you are making use of a RNG. There is one that is [specifically created for the ancient-map drawing](https://github.com/esampson/worldengine/blob/64f8b74d5a7c43e486f05f25a878819914d5a527/worldengine/drawing_functions.py#L280). You should hand it to ```_createmask()``` and use it (without reseeding it). Currently your RNG-output will be the same for every call of ```_createmask()``` since you reuse the same seed over and over. Might give odd results.

[draw.py - all over the place](https://github.com/Mindwerks/worldengine/compare/master...esampson:NewMap#diff-993da1fbf1c1e611672d9ced4ed7b4a7R409)
I love how much code you have gotten rid of. There was a lot of repetition in that file, I had already been thinking about doing some of the things you have done. I won't touch the file, hoping for your work to not die before it's finished. ;)

At least one of my PRs would break your code, if merged. Assuming I found out how, would you mind me sending you a PR to fix things up, in case it was necessary? I wouldn't want you to have to adapt your (great!) work to mine when it's probably a lot easier the other way around.

**Comment by esampson** [comment]
The purpose of the random number generation in _createmask() is to try and break up the placement of the symbols a bit. Previously a symbol would be placed at the very first spot of the territory, then the next would be placed absolute as soon as possible, etc. This had a tendency to create an extremely regular pattern. Using the RNG creates a mask that is more like a scattering of possible locations. Then when the symbols are filled in the first symbol is placed at the earliest possibility, the next is placed at the next possible location, but that in no longer as soon as possible, etc.

I thought about the reseeding and wondered if it would have a noticeable effect. After a quick test it seemed that the effect was negligible. The underlying pattern repeats but it is offset throughout the various areas because of where they first start, how long each 'line' is, and the fact that they have different 'thresholds'. 

Then that is further munged up by the fact that the various symbols are not all of the same size. As a result when a pine tree clears possible locations it clears a smaller area than a deciduous tree. As a result, even if the underlying masks were identical the different sized symbols would probably break down into unique patterns.

Given all of that I thought that the advantage of reseeding (reducing possibility of minor code changes changing the image output) outweighed the drawback.

**Comment by tcld** [comment]
In that case you could, for example, strictly use seed 0. It still wouldn't have a noticeable effect but the code will look more deliberate. A copy of your comment here would also be a good idea. (*I realize that I am criticizing work-in-progress code. I'll stop annoying you with that practice ASAP.* ;) )
Overall I agree that a couple pixels of variation, which won't really be visually noticeable, can be sacrificed for the sake of certain code-changes not changing the output. (Although once you start changing the code you probably want to change more than the order of the calls to ```_createmask()``` anyway, so the result might change regardless.)

Something else: Since you replaced all the mask-functions with a single one, how about replacing all the [```unset_mask()```-functions](https://github.com/Mindwerks/worldengine/compare/master...esampson:NewMap#diff-993da1fbf1c1e611672d9ced4ed7b4a7R338), too? I'm not sure how possible it is. To me ```unset_boreal_mask(pos)``` doesn't look any more clear than ```unset_mask(boreal_mask, pos)```. (You might want to place ```_createmask()``` inside of ```ancient_map()```, too, now that I think about it.)

**Comment by esampson** [comment]
I'm probably going to move both the _createmask and unset_mask functions into the MapComponent class (and I will possibly rename MapComponent to something else like TerrainType). I left the unset for a while because I was thinking in C++ terms that if I said unset_make(boreal_mask, pos) that I would pass the value of boreal_mask and not the reference to boreal_mask. I later realized my mistake, but since I am planning on embedding the mask routines into the class I just didn't get around to cleaning those up yet.

**Comment by ftomassetti** [comment]
I am really, really, really looking forward to use this :)

**Comment by esampson** [comment]
Sorry. I've just been stepping back from this a little while I think about some things. It looks like we are going to be making a change so that we have specific groups of mountain heights (hills, mountain, tall mountains, etc.). I'm planning on switching to that and using different images for the different heights once we do that (one concern of mine has been how the express the different set of rules that mountains were using so they could be rescaled).

I've also been looking at how to do things without using PIL. One a certain level it won't be too hard (the math behind the routines is pretty simple) but performance will take hit since the core for PIL is in C.

**Comment by psi29a** [comment]
I think that if this relies on images (binary data), it is time we go about using submodules. It will mean a bit of extra work, but at least we would have a clean repo and image/data with the rest. :)

Couldn't we load the image data into numpy and let numpy handle the maths for us to scale the image?

**Comment by tcld** [comment]
numpy does not provide every necessary function, for example it cannot do a gaussian blur; SciPy seems to have the necessary capabilities, as far as I know (it includes some routines for image-processing). It is a big module on its own, though.
